### PR TITLE
Add option to treat boundary limits for minim

### PR DESCRIPTION
**Pull request by dtnguyen2**
# Note

This PR adds the option on how the optimization ```minim``` will behave if the free parameter is beyond the limit.  The default parameter ```reflect``` is True (will reflect by an equal amount about the limit).  If ```reflect``` is set to False then model function will return ```DBL_MAX``` (~1e308).

### CODE REVIEW

**Comment by DougBurke** [comment]
@dtnguyen2 - if `reflect=False` and the minim optimiser hits a situation where you are at a parameter limit does the parameter tend to stay at that boundary, get "pushed away" from the boundary because the statistic value is likely to be very large thanks to the `DBL_MAX`, or is it hard to say what happens?

File: sherpa/optmethods/src/_saoopt.cc [hunk 1]
```diff
@@ -833,10 +835,11 @@ static PyObject* py_minim( PyObject* self, PyObject* args,
   PyObject* py_function=NULL;
   DoubleArray par, step, lb, ub;
   IntArray finalsimplex;
-  int verbose, maxnfev, nfev, iquad, ierr, initsimplex;
+  int reflect, verbose, maxnfev, nfev, iquad, ierr, initsimplex;
   double fval, ftol, simp;
 
-  if ( !PyArg_ParseTuple( args, (char*) "iiiiddO&O&O&O&O",
+  if ( !PyArg_ParseTuple( args, (char*) "iiiiiddO&O&O&O&O",
```
Comment (DougBurke) [inline sherpa/optmethods/src/_saoopt.cc:841]:
```suggestion
  if ( !PyArg_ParseTuple( args, (char*) "piiiiddO&O&O&O&O",
```

The p argument means the python value is treated as a boolean - see https://docs.python.org/3/c-api/arg.html#other-objects - and the result for `reflect` will be 0 or 1. It is possible that the second `i` could also be made a `p` as I assume `verbose` is also a boolean flag.

File: sherpa/optmethods/src/Simplex.hh [hunk 2]
```diff
@@ -276,7 +276,11 @@ public:
 
 private:
 
+#if (__STDC_VERSION__ < 201112L)
```
Comment (DougBurke) [inline sherpa/optmethods/src/Simplex.hh:279]: I know this check should work but I'm trying this on `sherpa/astro/utils/src/_pileup.cc` - i.e.

```
% git diff
diff --git a/sherpa/astro/utils/src/_pileup.cc b/sherpa/astro/utils/src/_pileup.cc
index 1cae067f..52ef947d 100644
--- a/sherpa/astro/utils/src/_pileup.cc
+++ b/sherpa/astro/utils/src/_pileup.cc
@@ -123,8 +123,13 @@ static PyObject* _apply_pileup( PyObject* self, PyObject* args )
     return NULL;
   }

+#if (__STDC_VERSION__ < 201112L)
   std::auto_ptr< sherpa::PyWrapper >
     wrapper( new sherpa::PyWrapper( py_function ) );
+#else
+  std::unique_ptr< sherpa::PyWrapper >
+    wrapper( new sherpa::PyWrapper( py_function ) );
+#endif

   DoubleArray results;
   if ( EXIT_SUCCESS != results.create( arf_source.get_ndim(),
```

and when I compile it with a relatively recent g++ it still complains, even with

    export CFLAGS=-std=c++11

we get

```
compiling C++ sources
C compiler: g++ -pthread -B /home/dburke/anaconda/envs/sherpa-master/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -std=c++11 -fPIC

compile options: '-Isherpa/include -Isherpa/utils/src -Isherpa/astro/utils/src -I/home/dburke/anaconda/envs/sherpa-master/lib/python3.8/site-packages/numpy/core/include -Ibuild/src.linux-x86_64-3.8/numpy/distutils/include -I/home/dburke/anaconda/envs/sherpa-master/include/python3.8 -I/home/dburke/anaconda/envs/sherpa-master/include/python3.8 -c'
extra options: '-msse -msse2 -msse3'
g++: sherpa/astro/utils/src/_pileup.cc
sherpa/astro/utils/src/_pileup.cc: In function ‘PyObject* _apply_pileup(PyObject*, PyObject*)’:
sherpa/astro/utils/src/_pileup.cc:127:8: warning: ‘template<class> class std::auto_ptr’ is deprecated [-Wdeprecated-declarations]
  127 |   std::auto_ptr< sherpa::PyWrapper >
      |        ^~~~~~~~
In file included from /usr/include/c++/10/memory:83,
                 from sherpa/astro/utils/src/_pileup.cc:22:
/usr/include/c++/10/bits/unique_ptr.h:57:28: note: declared here
   57 |   template<typename> class auto_ptr;
      |                            ^~~~~~~~
```

and line 127 is the #if-protected std::auto_ptr line

```
% cat -n sherpa/astro/utils/src/_pileup.cc
...
   125
   126  #if (__STDC_VERSION__ < 201112L)
   127    std::auto_ptr< sherpa::PyWrapper >
   128      wrapper( new sherpa::PyWrapper( py_function ) );
   129  #else
   130    std::unique_ptr< sherpa::PyWrapper >
   131      wrapper( new sherpa::PyWrapper( py_function ) );
   132  #endif
   133
..
```



For reference:

```
% g++ --version
g++ (Ubuntu 10.3.0-1ubuntu1) 10.3.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Comment (dtnguyen2) [reply sherpa/optmethods/src/Simplex.hh:279]: devel12-116: cat tmp.cc

```
#include <memory>
#include <iostream>

class Fubar {
public:
  Fubar() { }
};

int main() {
  std::cout << __cplusplus << '\n';
  Fubar* tmp = new Fubar();
#if (__cplusplus < 201112L)
  std::cout << "will use auto_ptr\n";
  std::auto_ptr< Fubar > myauto(tmp);
#else
  std::cout << "will use unique_ptr\n";
  std::unique_ptr< Fubar > myunique(tmp);
#endif
}
```
Using an ancient version:
```
devel12-119: g++ --version
g++ (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

devel12-120: g++ tmp.cc ; a.out
199711
will use auto_ptr
```
A more modern version:
```
devel12-172: g++ --version
g++ (GCC) 7.4.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

devel12-173: g++ tmp.cc; valgrind a.out
==17526== Memcheck, a memory error detector
==17526== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==17526== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==17526== Command: a.out
==17526== 
201402
will use unique_ptr
==17526== 
==17526== HEAP SUMMARY:
==17526==     in use at exit: 0 bytes in 0 blocks
==17526==   total heap usage: 2 allocs, 2 frees, 72,705 bytes allocated
==17526== 
==17526== All heap blocks were freed -- no leaks are possible
==17526== 
==17526== For counts of detected and suppressed errors, rerun with: -v
==17526== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

BTW, I need to nuke this PR and create a new one cause @Marie-Terrell told me that I'm working on the main branch.  Oops

**Comment by dtnguyen2** [comment]
@Marie-Terrell said that I'm working off main. Close this PR, an updated version is PR #1285.

**Pull request closed without merging**